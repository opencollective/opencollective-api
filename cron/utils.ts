import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
