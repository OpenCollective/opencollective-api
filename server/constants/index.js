import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, roles, math, channels, transactions };
