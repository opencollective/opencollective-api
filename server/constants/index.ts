import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import expenseTypes from './expense_type';
import frequencies from './frequencies';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, expenseTypes, roles, math, channels, transactions, frequencies };
