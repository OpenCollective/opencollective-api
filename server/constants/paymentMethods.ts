export enum PAYMENT_METHOD_SERVICE {
  PAYPAL = 'paypal',
  BRAINTREE = 'braintree',
  STRIPE = 'stripe',
  OPENCOLLECTIVE = 'opencollective',
  PREPAID = 'prepaid',
}
export const PAYMENT_METHOD_SERVICES = Object.values(PAYMENT_METHOD_SERVICE);

export enum PAYMENT_METHOD_TYPE {
  CREDITCARD = 'creditcard',
  PREPAID = 'prepaid',
  PAYMENT = 'payment',
  SUBSCRIPTION = 'subscription',
  COLLECTIVE = 'collective',
  HOST = 'host',
  ADAPTIVE = 'adaptive',
  PAYPAL = 'paypal', // Braintree service supports multiple payment methods including PayPal
  GIFT_CARD = 'giftcard',
  MANUAL = 'manual',
}
export const PAYMENT_METHOD_TYPES = Object.values(PAYMENT_METHOD_TYPE);
