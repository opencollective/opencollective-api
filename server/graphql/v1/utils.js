import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
