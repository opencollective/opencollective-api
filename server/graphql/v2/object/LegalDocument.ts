import config from 'config';
import { GraphQLInt, GraphQLNonNull, GraphQLObjectType, GraphQLString } from 'graphql';
import { GraphQLDateTime } from 'graphql-scalars';

import { GraphQLLegalDocumentRequestStatus } from '../enum/LegalDocumentRequestStatus';
import { GraphQLLegalDocumentService } from '../enum/LegalDocumentService';
import { GraphQLLegalDocumentType } from '../enum/LegalDocumentType';
import { getIdEncodeResolver, idEncode, IDENTIFIER_TYPES } from '../identifiers';
import { GraphQLAccount } from '../interface/Account';
import GraphQLURL from '../scalar/URL';

export const GraphQLLegalDocument = new GraphQLObjectType({
  name: 'LegalDocument',
  description: 'A legal document (e.g. W9, W8BEN, W8BEN-E)',
  fields: () => ({
    id: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'Unique identifier for this legal document',
      resolve: getIdEncodeResolver(IDENTIFIER_TYPES.LEGAL_DOCUMENT),
    },
    year: {
      type: new GraphQLNonNull(GraphQLInt),
      description: 'The year this legal document is for',
    },
    type: {
      type: new GraphQLNonNull(GraphQLLegalDocumentType),
      description: 'The type of legal document',
      resolve: ({ documentType }) => documentType,
    },
    status: {
      type: new GraphQLNonNull(GraphQLLegalDocumentRequestStatus),
      description: 'The status of the request for this legal document',
      resolve: ({ requestStatus }) => requestStatus,
    },
    service: {
      type: new GraphQLNonNull(GraphQLLegalDocumentService),
      description: 'The service that provided this legal document',
    },
    requestedAt: {
      type: new GraphQLNonNull(GraphQLDateTime),
      description: 'The date and time the request for this legal document was created',
      resolve: ({ createdAt }) => createdAt,
    },
    updatedAt: {
      type: new GraphQLNonNull(GraphQLDateTime),
      description: 'The date and time this legal document was last updated',
    },
    account: {
      type: new GraphQLNonNull(GraphQLAccount),
      description: 'The account this legal document is for',
    },
    documentLink: {
      type: GraphQLURL,
      description:
        'URL to download the file. Must be logged in as an account admin. The returned URL will be protected by authentication + 2FA.',
      resolve: async (document, _, req) => {
        if (document.documentLink && req.remoteUser) {
          const account = await req.loaders.Collective.byId.load(document.CollectiveId);
          if (req.remoteUser.isAdminOfCollective(account)) {
            return `${config.host.api}/legal-documents/${idEncode(document.id, 'legal-document')}/download`;
          }
        }
      },
    },
  }),
});
