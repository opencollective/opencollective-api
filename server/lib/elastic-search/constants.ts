export enum ElasticSearchIndexName {
  COLLECTIVES = 'collectives',
  COMMENTS = 'comments',
  EXPENSES = 'expenses',
  UPDATES = 'updates',
  TRANSACTIONS = 'transactions',
  ORDERS = 'orders',
  TIERS = 'tiers',
  HOST_APPLICATIONS = 'host-applications',
}
