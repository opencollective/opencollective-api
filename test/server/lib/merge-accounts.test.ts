import { expect } from 'chai';
import { mergeWith } from 'lodash';

import { PAYMENT_METHOD_SERVICE, PAYMENT_METHOD_TYPE } from '../../../server/constants/paymentMethods';
import { getMovableItemsCounts, mergeAccounts, simulateMergeAccounts } from '../../../server/lib/merge-accounts';
import models from '../../../server/models';
import { LEGAL_DOCUMENT_TYPE } from '../../../server/models/LegalDocument';
import { MigrationLogDataForMergeAccounts, MigrationLogType } from '../../../server/models/MigrationLog';
import { getOrCreateDBSnapshot } from '../../test-helpers/data-snapshot';
import * as Faker from '../../test-helpers/fake-data';
import { resetTestDB } from '../../utils';

/** Helper to create an account with many associations */
const addFakeDataToAccount = async (account): Promise<void> => {
  // Pre-generating a random collective to save some performance
  const randomCollective = await Faker.fakeCollective();
  const user = account.type === 'USER' ? await account.getUser() : await Faker.fakeUser();

  await Promise.all([
    Faker.fakeAccountingCategory({ CollectiveId: account.id }),
    Faker.fakeAgreement({ CollectiveId: account.id, UserId: user.id }),
    Faker.fakeActivity({ CollectiveId: account.id }, { hooks: false }),
    Faker.fakeApplication({ CollectiveId: account.id }),
    Faker.fakeComment({ CollectiveId: account.id, FromCollectiveId: randomCollective.id }, { hooks: false }),
    Faker.fakeComment({ FromCollectiveId: account.id, CollectiveId: randomCollective.id }, { hooks: false }),
    Faker.fakeConnectedAccount({ CollectiveId: account.id }, { hooks: false }),
    Faker.fakeConversation({ CollectiveId: account.id, FromCollectiveId: randomCollective.id }, { hooks: false }),
    Faker.fakeConversation({ FromCollectiveId: account.id, CollectiveId: randomCollective.id }, { hooks: false }),
    Faker.fakeEmojiReaction({ FromCollectiveId: account.id }),
    Faker.fakeExpense({ CollectiveId: account.id, FromCollectiveId: randomCollective.id, UserId: user.id }),
    Faker.fakeExpense({ FromCollectiveId: account.id, CollectiveId: randomCollective.id, UserId: user.id }),
    // TODO Faker.fakeHostApplication({ HostCollectiveId: account.id, CollectiveId: randomCollective.id }),
    // TODO Faker.fakeHostApplication({ CollectiveId: account.id, HostCollectiveId: randomCollective.id }),
    Faker.fakeLegalDocument({ CollectiveId: account.id, year: 2020, documentType: LEGAL_DOCUMENT_TYPE.US_TAX_FORM }),
    Faker.fakeLocation({ CollectiveId: account.id }),
    // TODO Faker.fakeMemberInvitation({ MemberCollectiveId: account.id }),
    Faker.fakeMember({ MemberCollectiveId: account.id, CollectiveId: randomCollective.id }),
    Faker.fakeNotification({ CollectiveId: account.id, UserId: user.id }),
    Faker.fakeOrder({
      FromCollectiveId: account.id,
      CollectiveId: randomCollective.id,
      CreatedByUserId: user.id,
    }),
    Faker.fakeOrder({
      CollectiveId: account.id,
      FromCollectiveId: randomCollective.id,
      CreatedByUserId: user.id,
    }),
    Faker.fakePaymentMethod({
      CollectiveId: account.id,
      service: PAYMENT_METHOD_SERVICE.STRIPE,
      type: PAYMENT_METHOD_TYPE.CREDITCARD,
    }),
    Faker.fakePayoutMethod({ CollectiveId: account.id }),
    Faker.fakePaypalProduct({ CollectiveId: account.id }),
    // TODO Faker.fakeRequiredLegalDocument({ HostCollectiveId: account.id }),
    // TODO: Add Oauth authorization codes
    Faker.fakeSocialLink({ CollectiveId: account.id }),
    Faker.fakeTier({ CollectiveId: account.id }),
    Faker.fakeTransaction({ FromCollectiveId: account.id, CollectiveId: randomCollective.id }),
    Faker.fakeTransaction({ CollectiveId: account.id, FromCollectiveId: randomCollective.id }),
    Faker.fakeTransaction({
      UsingGiftCardFromCollectiveId: account.id,
      FromCollectiveId: randomCollective.id,
      CollectiveId: randomCollective.id,
    }),
    Faker.fakeUpdate({ CollectiveId: account.id }, { hooks: false }),
    Faker.fakeUpdate({ FromCollectiveId: account.id }, { hooks: false }),
  ]);

  if (account.type !== 'USER') {
    await Promise.all([
      // TODO Faker.fakeMemberInvitation({ CollectiveId: account.id }),
      Faker.fakeMember({ CollectiveId: account.id, MemberCollectiveId: randomCollective.id }),
      Faker.fakeCollective({ HostCollectiveId: account.id }, { hooks: false }),
      Faker.fakeCollective({ ParentCollectiveId: account.id, HostCollectiveId: null }, { hooks: false }),
    ]);

    if (account.HostCollectiveId) {
      await Promise.all([
        Faker.fakeVirtualCard({ CollectiveId: account.id, HostCollectiveId: randomCollective.id }),
        Faker.fakeVirtualCard({ HostCollectiveId: account.id, CollectiveId: randomCollective.id }),
      ]);
    }
  }
};

/** Generate base data for the test. Remember to update the DB snapshot if you touch this function.  */
const generateTestData = async () => {
  // Make sure we're on a fresh DB
  await resetTestDB();

  // Create noise data to make sure merge tools don't affect others data
  await addFakeDataToAccount(await Faker.fakeCollective({ slug: 'noise-collective' }));
  await addFakeDataToAccount(await Faker.fakeOrganization({ slug: 'noise-organization' }));
  await addFakeDataToAccount((await Faker.fakeUser({}, { slug: 'noise-user' })).collective);

  // Create the accounts that will be merged
  const fromCollective = await Faker.fakeCollective({ slug: 'from-collective' });
  const toCollective = await Faker.fakeCollective({ slug: 'to-collective' });
  await addFakeDataToAccount(fromCollective);
  await addFakeDataToAccount(toCollective);

  const fromUser = await Faker.fakeUser({}, { slug: 'from-user', countryISO: 'FR' });
  const toUser = await Faker.fakeUser({}, { slug: 'to-user', countryISO: null });
  await addFakeDataToAccount(fromUser.collective);
  await addFakeDataToAccount(toUser.collective);

  const fromOrganization = await Faker.fakeCollective({ slug: 'from-org', countryISO: 'FR' });
  const toOrganization = await Faker.fakeCollective({ slug: 'to-org', countryISO: null });
  await addFakeDataToAccount(fromOrganization);
  await addFakeDataToAccount(toOrganization);
};

const sumCounts = (count1, count2) => {
  return {
    account: mergeWith(count1.account, count2.account, (objValue, srcValue, key) => {
      const sum = (objValue || 0) + (srcValue || 0);
      // Locations are merged in a different way to only keep the most recent one
      if (key === 'location') {
        return Math.min(1, sum);
      }
      return sum;
    }),
    user: !count1.user
      ? null
      : mergeWith(count1.user, count2.user, (objValue, srcValue) => {
          return (objValue || 0) + (srcValue || 0);
        }),
  };
};

describe('server/lib/merge-accounts', () => {
  let fromUser, toUser, fromOrganization, toOrganization, fromCollective, toCollective;

  before(async function () {
    await getOrCreateDBSnapshot(this, 'merge-accounts', generateTestData);

    // Load accounts
    fromCollective = await models.Collective.findBySlug('from-collective');
    toCollective = await models.Collective.findBySlug('to-collective');
    fromOrganization = await models.Collective.findBySlug('from-org');
    toOrganization = await models.Collective.findBySlug('to-org');

    // Load users
    const fromUserCollective = await models.Collective.findBySlug('from-user');
    const toUserCollective = await models.Collective.findBySlug('to-user');
    fromUser = await models.User.findOne({ where: { CollectiveId: fromUserCollective.id } });
    toUser = await models.User.findOne({ where: { CollectiveId: toUserCollective.id } });
    fromUser.collective = fromUserCollective;
    toUser.collective = toUserCollective;
  });

  describe('simulateMergeAccounts', () => {
    it('Correctly estimates the number of items to move for collective account', async () => {
      // Generate & check summary
      const summary = await simulateMergeAccounts(fromCollective, toCollective);
      expect(summary).to.matchSnapshot();
    });

    it('Correctly estimates the number of items to move for organization account', async () => {
      // Generate & check summary
      const summary = await simulateMergeAccounts(fromOrganization, toOrganization);
      expect(summary).to.matchSnapshot();
    });

    it('Correctly estimates the number of items to move for user account', async () => {
      // Generate & check summary
      const summary = await simulateMergeAccounts(fromUser.collective, toUser.collective);
      expect(summary).to.matchSnapshot();
    });
  });

  describe('mergeAccounts', () => {
    it('Merges an organization', async () => {
      // Check seed data
      const preMoveFromItemsCounts = await getMovableItemsCounts(fromOrganization);
      const preMoveToItemsCounts = await getMovableItemsCounts(toOrganization);
      expect(preMoveFromItemsCounts.account).to.matchSnapshot();
      expect(preMoveToItemsCounts.account).to.matchSnapshot();
      expect(preMoveFromItemsCounts.user).to.be.null;
      expect(preMoveToItemsCounts.user).to.be.null;

      // Merge accounts
      await mergeAccounts(fromOrganization, toOrganization);

      // Profile info
      await fromOrganization.reload({ paranoid: false });
      await toOrganization.reload();
      expect(fromOrganization.deletedAt).to.not.be.null;
      expect(fromOrganization.slug).to.eq('from-org-merged');
      expect(fromOrganization.data.mergedIntoCollectiveId).to.eq(toOrganization.id);
      expect(toOrganization.countryISO).to.eq('FR');

      // Associated data
      const postMoveFromItemsCounts = await getMovableItemsCounts(fromOrganization);
      const postMoveToItemsCounts = await getMovableItemsCounts(toOrganization);
      const expectedCounts = sumCounts(preMoveFromItemsCounts, preMoveToItemsCounts);
      expectedCounts.account.legalDocuments -= 1; // Should not be transferred as one already exists
      expect(postMoveToItemsCounts.account).to.matchSnapshot();
      expect(postMoveToItemsCounts).to.deep.equal(expectedCounts);
      expect(postMoveFromItemsCounts.user).to.be.null;
      expect(postMoveToItemsCounts.user).to.be.null;
      Object.values(postMoveFromItemsCounts.account).forEach(count => expect(count).to.eq(0));

      // Creates a MigrationLog
      const migrationLog = await models.MigrationLog.findOne({
        where: {
          type: MigrationLogType.MERGE_ACCOUNTS,
          description: 'Merge from-org into to-org',
        },
      });

      const migrationLogData = <MigrationLogDataForMergeAccounts>migrationLog.data;
      expect(migrationLogData).to.exist;
      expect(migrationLogData.fromAccount).to.eq(fromOrganization.id);
      expect(migrationLogData.intoAccount).to.eq(toOrganization.id);
      expect(migrationLogData.associations.members).to.have.length(2);
      expect(migrationLogData.associations.expenses).to.have.length(1);
      expect(migrationLogData.associations.giftCardTransactions).to.have.length(1);
      expect(migrationLogData.associations.socialLinks).to.have.length(1);
      expect(migrationLogData.associations.socialLinks[0]['CollectiveId']).to.eq(toOrganization.id);
      expect(migrationLogData.associations.socialLinks[0]['type']).to.exist;
      expect(migrationLogData.associations.socialLinks[0]['url']).to.exist;
    });

    it('Merges a user profile', async () => {
      // Check seed data
      const preMoveFromItemsCounts = await getMovableItemsCounts(fromUser.collective);
      const preMoveToItemsCounts = await getMovableItemsCounts(toUser.collective);
      expect(preMoveFromItemsCounts.account).to.matchSnapshot();
      expect(preMoveToItemsCounts.account).to.matchSnapshot();
      expect(preMoveFromItemsCounts.user).to.matchSnapshot();
      expect(preMoveToItemsCounts.user).to.matchSnapshot();

      // Prepare test data
      await mergeAccounts(fromUser.collective, toUser.collective);

      // Profile info
      await fromUser.reload({ paranoid: false });
      await fromUser.collective.reload({ paranoid: false });
      await toUser.reload();
      expect(fromUser.deletedAt).to.not.be.null;
      expect(fromUser.collective.deletedAt).to.not.be.null;
      expect(fromUser.collective.slug).to.eq('from-user-merged');
      expect(fromUser.collective.data.mergedIntoCollectiveId).to.eq(toUser.CollectiveId);
      expect(fromUser.data.mergedIntoUserId).to.eq(toUser.id);
      expect(toUser.collective.countryISO).to.eq('FR');

      // Associated data
      const postMoveFromItemsCounts = await getMovableItemsCounts(fromUser.collective);
      const postMoveToItemsCounts = await getMovableItemsCounts(toUser.collective);
      const expectedCounts = sumCounts(preMoveFromItemsCounts, preMoveToItemsCounts);
      expectedCounts.account.legalDocuments -= 1; // Should not be transferred as one already exists
      expectedCounts.user.collectives -= 1; // User profile is not merged, not transferred
      expect(postMoveToItemsCounts.account).to.matchSnapshot();
      expect(postMoveToItemsCounts).to.deep.equal(expectedCounts);
      Object.values(postMoveFromItemsCounts.account).forEach(count => expect(count).to.eq(0));
      Object.values(postMoveFromItemsCounts.user).forEach(count => expect(count).to.eq(0));

      // Creates a MigrationLog
      const migrationLog = await models.MigrationLog.findOne({
        where: {
          type: MigrationLogType.MERGE_ACCOUNTS,
          description: 'Merge from-user into to-user',
        },
      });

      const migrationLogData = <MigrationLogDataForMergeAccounts>migrationLog.data;
      expect(migrationLogData).to.exist;
      expect(migrationLogData.fromAccount).to.eq(fromUser.collective.id);
      expect(migrationLogData.intoAccount).to.eq(toUser.collective.id);
      expect(migrationLogData.associations.giftCardTransactions).to.have.length(1);
      expect(migrationLogData.associations.expenses).to.have.length(1);
      expect(migrationLogData.userChanges.notifications).to.have.length(1);
    });
  });
});
