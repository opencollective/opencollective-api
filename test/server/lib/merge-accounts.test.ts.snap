// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server/lib/merge-accounts mergeAccounts Merges a user profile 1`] = `
Object {
  "accountingCategories": 1,
  "activities": 1,
  "agreements": 1,
  "applications": 1,
  "childrenCollectives": 0,
  "comments": 1,
  "commentsCreated": 1,
  "connectedAccounts": 1,
  "conversations": 1,
  "conversationsCreated": 1,
  "creditTransactions": 1,
  "debitTransactions": 1,
  "emojiReactions": 1,
  "expenses": 1,
  "expensesCreated": 1,
  "giftCardTransactions": 1,
  "hostApplications": 0,
  "hostApplicationsCreated": 0,
  "hostedAgreements": 0,
  "hostedCollectives": 0,
  "hostedVirtualCardRequests": 0,
  "legalDocuments": 1,
  "location": 1,
  "memberInvitations": 0,
  "members": 1,
  "membershipInvitations": 0,
  "memberships": 0,
  "notifications": 1,
  "ordersCreated": 1,
  "ordersReceived": 1,
  "paymentMethods": 1,
  "payoutMethods": 1,
  "paypalProducts": 1,
  "personalTokens": 0,
  "recurringExpenses": 0,
  "recurringExpensesCreated": 0,
  "requiredLegalDocuments": 0,
  "socialLinks": 1,
  "tiers": 1,
  "updates": 1,
  "updatesCreated": 1,
  "virtualCardRequests": 0,
  "virtualCards": 0,
  "virtualCardsHosted": 0,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges a user profile 2`] = `
Object {
  "accountingCategories": 1,
  "activities": 1,
  "agreements": 1,
  "applications": 1,
  "childrenCollectives": 0,
  "comments": 1,
  "commentsCreated": 1,
  "connectedAccounts": 1,
  "conversations": 1,
  "conversationsCreated": 1,
  "creditTransactions": 1,
  "debitTransactions": 1,
  "emojiReactions": 1,
  "expenses": 1,
  "expensesCreated": 1,
  "giftCardTransactions": 1,
  "hostApplications": 0,
  "hostApplicationsCreated": 0,
  "hostedAgreements": 0,
  "hostedCollectives": 0,
  "hostedVirtualCardRequests": 0,
  "legalDocuments": 1,
  "location": 1,
  "memberInvitations": 0,
  "members": 1,
  "membershipInvitations": 0,
  "memberships": 0,
  "notifications": 1,
  "ordersCreated": 1,
  "ordersReceived": 1,
  "paymentMethods": 1,
  "payoutMethods": 1,
  "paypalProducts": 1,
  "personalTokens": 0,
  "recurringExpenses": 0,
  "recurringExpensesCreated": 0,
  "requiredLegalDocuments": 0,
  "socialLinks": 1,
  "tiers": 1,
  "updates": 1,
  "updatesCreated": 1,
  "virtualCardRequests": 0,
  "virtualCards": 0,
  "virtualCardsHosted": 0,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges a user profile 3`] = `
Object {
  "activities": 0,
  "agreements": 1,
  "applications": 0,
  "collectives": 1,
  "comments": 0,
  "conversationFollowers": 0,
  "conversations": 0,
  "emojiReactions": 0,
  "expenseAttachedFiles": 0,
  "expenseItems": 0,
  "expenses": 2,
  "memberInvitations": 0,
  "members": 0,
  "migrationLogs": 0,
  "notifications": 1,
  "oAuthAuthorizationCodes": 0,
  "orders": 2,
  "paymentMethods": 0,
  "payoutMethods": 0,
  "personalTokens": 0,
  "transactions": 0,
  "twoFactorMethods": 0,
  "updates": 0,
  "uploadedFiles": 0,
  "userTokens": 0,
  "virtualCardRequests": 0,
  "virtualCards": 0,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges a user profile 4`] = `
Object {
  "activities": 0,
  "agreements": 1,
  "applications": 0,
  "collectives": 1,
  "comments": 0,
  "conversationFollowers": 0,
  "conversations": 0,
  "emojiReactions": 0,
  "expenseAttachedFiles": 0,
  "expenseItems": 0,
  "expenses": 2,
  "memberInvitations": 0,
  "members": 0,
  "migrationLogs": 0,
  "notifications": 1,
  "oAuthAuthorizationCodes": 0,
  "orders": 2,
  "paymentMethods": 0,
  "payoutMethods": 0,
  "personalTokens": 0,
  "transactions": 0,
  "twoFactorMethods": 0,
  "updates": 0,
  "uploadedFiles": 0,
  "userTokens": 0,
  "virtualCardRequests": 0,
  "virtualCards": 0,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges a user profile 5`] = `
Object {
  "accountingCategories": 2,
  "activities": 2,
  "agreements": 2,
  "applications": 2,
  "childrenCollectives": 0,
  "comments": 2,
  "commentsCreated": 2,
  "connectedAccounts": 2,
  "conversations": 2,
  "conversationsCreated": 2,
  "creditTransactions": 2,
  "debitTransactions": 2,
  "emojiReactions": 2,
  "expenses": 2,
  "expensesCreated": 2,
  "giftCardTransactions": 2,
  "hostApplications": 0,
  "hostApplicationsCreated": 0,
  "hostedAgreements": 0,
  "hostedCollectives": 0,
  "hostedVirtualCardRequests": 0,
  "legalDocuments": 1,
  "location": 1,
  "memberInvitations": 0,
  "members": 2,
  "membershipInvitations": 0,
  "memberships": 0,
  "notifications": 2,
  "ordersCreated": 2,
  "ordersReceived": 2,
  "paymentMethods": 2,
  "payoutMethods": 2,
  "paypalProducts": 2,
  "personalTokens": 0,
  "recurringExpenses": 0,
  "recurringExpensesCreated": 0,
  "requiredLegalDocuments": 0,
  "socialLinks": 2,
  "tiers": 2,
  "updates": 2,
  "updatesCreated": 2,
  "virtualCardRequests": 0,
  "virtualCards": 0,
  "virtualCardsHosted": 0,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges an organization 1`] = `
Object {
  "accountingCategories": 1,
  "activities": 1,
  "agreements": 1,
  "applications": 1,
  "childrenCollectives": 1,
  "comments": 1,
  "commentsCreated": 1,
  "connectedAccounts": 1,
  "conversations": 1,
  "conversationsCreated": 1,
  "creditTransactions": 1,
  "debitTransactions": 1,
  "emojiReactions": 1,
  "expenses": 1,
  "expensesCreated": 1,
  "giftCardTransactions": 1,
  "hostApplications": 0,
  "hostApplicationsCreated": 0,
  "hostedAgreements": 0,
  "hostedCollectives": 1,
  "hostedVirtualCardRequests": 0,
  "legalDocuments": 1,
  "location": 1,
  "memberInvitations": 0,
  "members": 2,
  "membershipInvitations": 0,
  "memberships": 2,
  "notifications": 1,
  "ordersCreated": 1,
  "ordersReceived": 1,
  "paymentMethods": 2,
  "payoutMethods": 1,
  "paypalProducts": 1,
  "personalTokens": 0,
  "recurringExpenses": 0,
  "recurringExpensesCreated": 0,
  "requiredLegalDocuments": 0,
  "socialLinks": 1,
  "tiers": 1,
  "updates": 1,
  "updatesCreated": 1,
  "virtualCardRequests": 0,
  "virtualCards": 1,
  "virtualCardsHosted": 1,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges an organization 2`] = `
Object {
  "accountingCategories": 1,
  "activities": 1,
  "agreements": 1,
  "applications": 1,
  "childrenCollectives": 1,
  "comments": 1,
  "commentsCreated": 1,
  "connectedAccounts": 1,
  "conversations": 1,
  "conversationsCreated": 1,
  "creditTransactions": 1,
  "debitTransactions": 1,
  "emojiReactions": 1,
  "expenses": 1,
  "expensesCreated": 1,
  "giftCardTransactions": 1,
  "hostApplications": 0,
  "hostApplicationsCreated": 0,
  "hostedAgreements": 0,
  "hostedCollectives": 1,
  "hostedVirtualCardRequests": 0,
  "legalDocuments": 1,
  "location": 1,
  "memberInvitations": 0,
  "members": 2,
  "membershipInvitations": 0,
  "memberships": 2,
  "notifications": 1,
  "ordersCreated": 1,
  "ordersReceived": 1,
  "paymentMethods": 2,
  "payoutMethods": 1,
  "paypalProducts": 1,
  "personalTokens": 0,
  "recurringExpenses": 0,
  "recurringExpensesCreated": 0,
  "requiredLegalDocuments": 0,
  "socialLinks": 1,
  "tiers": 1,
  "updates": 1,
  "updatesCreated": 1,
  "virtualCardRequests": 0,
  "virtualCards": 1,
  "virtualCardsHosted": 1,
}
`;

exports[`server/lib/merge-accounts mergeAccounts Merges an organization 3`] = `
Object {
  "accountingCategories": 2,
  "activities": 2,
  "agreements": 2,
  "applications": 2,
  "childrenCollectives": 2,
  "comments": 2,
  "commentsCreated": 2,
  "connectedAccounts": 2,
  "conversations": 2,
  "conversationsCreated": 2,
  "creditTransactions": 2,
  "debitTransactions": 2,
  "emojiReactions": 2,
  "expenses": 2,
  "expensesCreated": 2,
  "giftCardTransactions": 2,
  "hostApplications": 0,
  "hostApplicationsCreated": 0,
  "hostedAgreements": 0,
  "hostedCollectives": 2,
  "hostedVirtualCardRequests": 0,
  "legalDocuments": 1,
  "location": 1,
  "memberInvitations": 0,
  "members": 4,
  "membershipInvitations": 0,
  "memberships": 4,
  "notifications": 2,
  "ordersCreated": 2,
  "ordersReceived": 2,
  "paymentMethods": 4,
  "payoutMethods": 2,
  "paypalProducts": 2,
  "personalTokens": 0,
  "recurringExpenses": 0,
  "recurringExpensesCreated": 0,
  "requiredLegalDocuments": 0,
  "socialLinks": 2,
  "tiers": 2,
  "updates": 2,
  "updatesCreated": 2,
  "virtualCardRequests": 0,
  "virtualCards": 2,
  "virtualCardsHosted": 2,
}
`;

exports[`server/lib/merge-accounts simulateMergeAccounts Correctly estimates the number of items to move for collective account 1`] = `
"The profiles information will be merged.

The following items will be moved to @to-collective:
  - accountingCategories: 1
  - agreements: 1
  - activities: 1
  - applications: 1
  - childrenCollectives: 1
  - comments: 1
  - commentsCreated: 1
  - connectedAccounts: 1
  - conversations: 1
  - conversationsCreated: 1
  - creditTransactions: 1
  - debitTransactions: 1
  - emojiReactions: 1
  - expenses: 1
  - expensesCreated: 1
  - giftCardTransactions: 1
  - hostedCollectives: 1
  - legalDocuments: 1
  - location: 1
  - members: 2
  - memberships: 2
  - notifications: 1
  - ordersCreated: 1
  - ordersReceived: 1
  - paymentMethods: 2
  - payoutMethods: 1
  - paypalProducts: 1
  - socialLinks: 1
  - tiers: 1
  - updates: 1
  - updatesCreated: 1
  - virtualCards: 1
  - virtualCardsHosted: 1

"
`;

exports[`server/lib/merge-accounts simulateMergeAccounts Correctly estimates the number of items to move for organization account 1`] = `
"The profiles information will be merged.

The following items will be moved to @to-org:
  - accountingCategories: 1
  - agreements: 1
  - activities: 1
  - applications: 1
  - childrenCollectives: 1
  - comments: 1
  - commentsCreated: 1
  - connectedAccounts: 1
  - conversations: 1
  - conversationsCreated: 1
  - creditTransactions: 1
  - debitTransactions: 1
  - emojiReactions: 1
  - expenses: 1
  - expensesCreated: 1
  - giftCardTransactions: 1
  - hostedCollectives: 1
  - legalDocuments: 1
  - location: 1
  - members: 2
  - memberships: 2
  - notifications: 1
  - ordersCreated: 1
  - ordersReceived: 1
  - paymentMethods: 2
  - payoutMethods: 1
  - paypalProducts: 1
  - socialLinks: 1
  - tiers: 1
  - updates: 1
  - updatesCreated: 1
  - virtualCards: 1
  - virtualCardsHosted: 1

"
`;

exports[`server/lib/merge-accounts simulateMergeAccounts Correctly estimates the number of items to move for user account 1`] = `
"The profiles information will be merged.

The following items will be moved to @to-user:
  - accountingCategories: 1
  - agreements: 1
  - activities: 1
  - applications: 1
  - comments: 1
  - commentsCreated: 1
  - connectedAccounts: 1
  - conversations: 1
  - conversationsCreated: 1
  - creditTransactions: 1
  - debitTransactions: 1
  - emojiReactions: 1
  - expenses: 1
  - expensesCreated: 1
  - giftCardTransactions: 1
  - legalDocuments: 1
  - location: 1
  - members: 1
  - notifications: 1
  - ordersCreated: 1
  - ordersReceived: 1
  - paymentMethods: 1
  - payoutMethods: 1
  - paypalProducts: 1
  - socialLinks: 1
  - tiers: 1
  - updates: 1
  - updatesCreated: 1

"
`;
